lib/OpenQA/WebAPI/Plugin/AMQP.pm

```diff
@@ -69,6 +70,178 @@ sub log_event {
+            $stdevent = 'error';
+        }
+        else {
+            $stdevent = 'complete';
+        }
+    }
+    else {
+        return;
+    }
+
+    # we need this for the system dict; it should be the release of
+    # the system-under-test (the VM in which the test runs) at the
+    # *start* of the test, I think. We're trying to capture info about
+    # the environment in which the test runs
+    my $sysrelease = $job->VERSION;
+    my $hdd1;
+    my $bootfrom;
+    $hdd1       = $job->settings_hash->{HDD_1}    if ($job->settings_hash->{HDD_1});
+    $bootfrom   = $job->settings_hash->{BOOTFROM} if ($job->settings_hash->{BOOTFROM});
+    $sysrelease = $1                              if ($hdd1 && $hdd1 =~ /disk_f(\d+)/ && $bootfrom eq 'c');
+
+    # next, get the 'artifact' (type of thing we tested)
+    my $artifact;
+    my $artifact_id;
+    my $artifact_release;
+    my $compose_type;
+    my $test_namespace;
+    # current date/time in ISO 8601 format
+    my $generated_at = strftime("%Y-%m-%dT%H:%M:%S", gmtime()) . 'Z';
+
+    my $build = $job->BUILD;
+    if ($build =~ /^Fedora/) {
+        $artifact       = 'productmd-compose';
+        $artifact_id    = $build;
+        $compose_type   = 'production';
+        $compose_type   = 'nightly' if ($build =~ /\.n\./);
+        $compose_type   = 'test' if ($build =~ /\.t\./);
+        $test_namespace = 'compose';
+    }
+    elsif ($build =~ /^Update-/) {
+        $artifact    = 'fedora-update';
+        $artifact_id = $build;
+        $artifact_id =~ s/^Update-//;
+        $artifact_release = $job->VERSION;
+        $test_namespace   = 'update';
+    }
+    else {
+        # unhandled artifact type
+        return;
+    }
+
+    # finally, construct the message content
+    my %msg_data = (
+        contact => {
+            name  => 'Fedora openQA',
+            team  => 'Fedora QA',
+            url   => $baseurl,
+            docs  => 'https://fedoraproject.org/wiki/OpenQA',
+            irc   => '#fedora-qa',
+            email => 'qa-devel@lists.fedoraproject.org',
+        },
+        run => {
+            url => "$baseurl/tests/$job_id",
+            log => "$baseurl/tests/$job_id/file/autoinst-log.txt",
+            id  => $job_id,
+        },
+        artifact => {
+            type => $artifact,
+            id   => $artifact_id,
+        },
+        pipeline => {
+            # per https://pagure.io/fedora-ci/messages/issue/61 this
+            # is meant to be unique per test scenario *and* artifact,
+            # so we construct it out of BUILD and the scenario keys.
+            # 'name' is supposed to be a 'human readable name', well,
+            # this is human readable, so we'll just use it twice
+            id   => join('.', "openqa", $job->BUILD, $job->TEST, $job->MACHINE, $job->FLAVOR, $job->ARCH),
+            name => join('.', "openqa", $job->BUILD, $job->TEST, $job->MACHINE, $job->FLAVOR, $job->ARCH),
```

Comment:
better to extract that into a variable then I guess

Reply:
eh, I guess, since they're literally right next to each other it didn't feel too bad.

Reply:
We try to avoid unnecessary sub calls in openQA now.